src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,235 @@
+of the `module k V`. -/
+structure affine_subspace :=
+(carrier : set P)
+(direction : submodule k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction)
+
+instance : has_coe (affine_subspace k V P) (set P) := ⟨affine_subspace.carrier⟩
+instance : has_mem P (affine_subspace k V P) := ⟨λ p s, p ∈ (s : set P)⟩
+
+/-- A point is in an affine subspace coerced to a set if and only if
+it is in that affine subspace. -/
+@[simp] lemma affine_subspace.mem_coe (p : P) (s : affine_subspace k V P) :
+  p ∈ (s : set P) ↔ p ∈ s :=
+iff.rfl
+
+/-- The whole affine space as a subspace of itself. -/
+def affine_subspace.univ : affine_subspace k V P :=
+{ carrier := set.univ,
+  direction := submodule.span k set.univ,
+  nonempty := set.nonempty_iff_univ_nonempty.1 S.nonempty,
+  add := λ p v hp hv, set.mem_univ _,
+  sub := begin
+    intros p1 p2 hp1 hp2,
+    apply set.mem_bInter,
+    intros x hx,
+    rw set.mem_set_of_eq at hx,
+    exact set.mem_of_mem_of_subset (set.mem_univ _) hx
+  end }
+
+/-- `affine_subspace.univ`, coerced to a set, is the whole set of
+points. -/
+@[simp] lemma affine_subspace.univ_coe : (affine_subspace.univ k V P : set P) = set.univ :=
+rfl
+
+/-- All points are in `affine_subspace.univ`. -/
+lemma affine_subspace.mem_univ (p : P) : p ∈ affine_subspace.univ k V P :=
+set.mem_univ p
+
+instance : inhabited (affine_subspace k V P) := ⟨affine_subspace.univ k V P⟩
+
+/-- The affine span of a nonempty set of points is the smallest affine
+subspace containing those points. (Actually defined here in terms of
+spans in modules.) -/
+def affine_span (s : set P) (h : s.nonempty) : affine_subspace k V P :=
+{ carrier := span_points k V s,
+  direction := vector_span k V s,
+  nonempty := span_points_nonempty_of_nonempty k V h,
+  add := λ p v hp hv, vadd_mem_span_points_of_mem_span_points_of_mem_vector_span k V hp hv,
+  sub := λ p1 p2 hp1 hp2, vsub_mem_vector_span_of_mem_span_points_of_mem_span_points k V hp1 hp2 }
+
+/-- The affine span, converted to a set, is `span_points`. -/
+@[simp] lemma affine_span_coe (s : set P) (h : s.nonempty) :
+  (affine_span k V P s h : set P) = span_points k V s :=
+rfl
+
+/-- A point in a set is in its affine span. -/
+lemma affine_span_mem (p : P) (s : set P) (hp : p ∈ s) : p ∈ affine_span k V P s ⟨p, hp⟩ :=
+mem_span_points k V p s hp
+
+end affine_subspace
+
+section affine_map
+
+variables  (k : Type*) (V1 : Type*) (P1 : Type*) (V2 : Type*) (P2 : Type*)
+    (V3 : Type*) (P3 : Type*) [ring k]
+    [add_comm_group V1] [module k V1] [affine_space k V1 P1]
+    [add_comm_group V2] [module k V2] [affine_space k V2 P2]
+    [add_comm_group V3] [module k V3] [affine_space k V3 P3]
+
+/-- An `affine_map k V1 P1 V2 P2` is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure affine_map :=
```

Comment:
It would be nice to have an `@[ext]` lemma here, stating that to check two affine maps are equal, it suffices to check that the coercions to functions agree. (And in particular that one doesn't need to compare the `linear` fields.)

Reply:
Added `affine_map.ext`.